src/puppetlabs/kitchensink/core.clj

```diff
@@ -1068,10 +1068,24 @@ to be a zipper."
        ~g)))
 
 (defn open-port-num
-  "Returns a currently open port number"
+  "Returns a currently open port number in the traditional ephemeral port range
+  of 49152 through 65535."
   []
-  (with-open [s (java.net.ServerSocket. 0)]
-    (.getLocalPort s)))
+  (loop []
+    (let [lo 49152
+          hi 65536 ; one higher because the upper limit is never selected
+          !the-port (atom nil)] ; because `catch` is not tail position
+      (try
+        (with-open [s (java.net.ServerSocket. (-> (rand-int (- hi lo)) (+ lo)))]
+          ;; have to use the atom to communicate the port (or lack thereof) to
+          ;; tail position, outside the `try` form
+          (reset! !the-port (.getLocalPort s)))
+        (catch java.net.BindException e
+          (when-not (re-find #"already in use" (.getMessage e))
+            (throw e))))
+      (if-let [the-port @!the-port]
+        the-port
+        (recur)))))
 
```

Comment:
```clojure
(defn- open-port?
  [port]
  (try
    (with-open [_ (java.net.ServerSocket. port)]
      port)
    (catch java.net.BindException e
      (when-not (re-find #"already in use" (.getMessage e))
        (throw e)))))

(defn open-port-num
  "Returns a currently open port number in the traditional ephemeral port range
  of 49152 through 65535."
  []
  (let [lo 49152
        hi 65536 ;; one higher because the upper limit is never selected
        ports (repeatedly #(-> (rand-int (- hi lo)) (+ lo)))]
    (some open-port? ports)))
```

my quick attempt at not doing atoms and manual looping